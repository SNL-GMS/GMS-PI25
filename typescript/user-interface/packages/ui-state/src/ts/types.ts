import type {
  ChannelSegmentTypes,
  ChannelTypes,
  CommonTypes,
  EventTypes,
  FilterTypes,
  FkTypes,
  SignalDetectionTypes
} from '@gms/common-model';
import type { AssociationConflict, Event } from '@gms/common-model/lib/event';
import type {
  FilterDefinition,
  FilterDefinitionByFilterDefinitionUsage
} from '@gms/common-model/lib/filter/types';
import type { QcSegment } from '@gms/common-model/lib/qc-segment';
import type { SignalDetectionHypothesis } from '@gms/common-model/lib/signal-detection';
import type { Station } from '@gms/common-model/lib/station-definitions/station-definitions/station-definitions';

export interface UiChannelSegment {
  channelSegmentDescriptor: ChannelSegmentTypes.ChannelSegmentDescriptor;
  processingMasks: ChannelSegmentTypes.ProcessingMask[];
  channelSegment: ChannelSegmentTypes.ChannelSegment<ChannelSegmentTypes.TimeSeries>;
  domainTimeRange?: CommonTypes.TimeRange;
}

export type ChannelName = UiChannelSegment['channelSegmentDescriptor']['channel']['name'];
export type StationName = Station['name'];
/** Weavess row could be a channel name or station name */
export type WeavessRowName = ChannelName | StationName;
export type FilterName = FilterDefinition['name'];
/**
 * A combination of the namedFilter name (if it exists) + filter definition name
 */
export type FilterNameId = string;
export type SampleRate = FilterDefinition['filterDescription']['parameters']['sampleRateHz'];
/**
 * JSON.stringify({@link ChannelSegmentTypes.ChannelSegmentDescriptor})
 */
export type ChannelSegmentDescriptorId = string;
export type SignalDetectionIdString = SignalDetectionTypes.SignalDetection['id'];
export type SignalDetectionHypothesisId = SignalDetectionHypothesis['id']['id'];
export type EventId = Event['id'];
export type QcSegmentId = QcSegment['id'];

/**
 * Record<{@link ChannelName}, Record<{@link FilterName}, {@link UiChannelSegment}[]>>
 *
 * A record of channel names, each containing a record of filter names to an array of ui channel segments.
 */
export type UIChannelSegmentRecord = Record<ChannelName, Record<FilterName, UiChannelSegment[]>>;

/**
 * Record<JSON.stringify({@link ChannelSegmentTypes.ChannelSegmentDescriptor}), {@link ChannelSegment<FkTypes.FkPowerSpectra>}>>
 *
 * A record of Fk Channel Segments
 */
export type FkChannelSegmentRecord = Record<
  ChannelSegmentDescriptorId,
  ChannelSegmentTypes.ChannelSegment<FkTypes.FkPowerSpectra>
>;

/**
 * Record<JSON.stringify({@link SignalDetectionIdString}), {@link FkTypes.FkFrequencyThumbnail[]}>>
 *
 * A record of FkFrequencyThumbnails associated to a SignalDetection
 */
export type FkFrequencyThumbnailRecord = Record<
  SignalDetectionIdString,
  FkTypes.FkFrequencyThumbnail[]
>;

/**
 * Record<{@link ChannelName}, {@link ChannelTypes.Channel}>
 *
 * A record of channel names to a matching channel.
 */
export type ChannelRecord = Record<ChannelName, ChannelTypes.Channel>;

/**
 * Record<{@link WeavessRowName}, {@link FilterTypes.Filter}>
 *
 * A record of weavess row names (either a channel name or station name) to a filter.
 */
export type ChannelFilterRecord = Record<WeavessRowName, FilterTypes.Filter>;

/**
 * Record<{@link FilterName}, Record<{@link SampleRate}, {@link FilterDefinition}[]>>
 *
 * A record of filter names, each containing a record of sample rates to a filter definition.
 */
export type FilterDefinitionsRecord = Record<FilterName, Record<SampleRate, FilterDefinition>>;

/**
 * Record<{@link SignalDetectionHypothesisId}, {@link FilterDefinitionByFilterDefinitionUsage}>
 */
export type FilterDefinitionsForSignalDetectionsRecord = Record<
  SignalDetectionHypothesisId,
  FilterDefinitionByFilterDefinitionUsage
>;

/**
 * Record<JSON.stringify({@link ChannelSegmentTypes.ChannelSegmentDescriptor}), {@link FilterDefinitionByFilterDefinitionUsage}>
 */
export type FilterDefinitionsForChannelSegmentsRecord = Record<
  ChannelSegmentDescriptorId,
  FilterDefinitionByFilterDefinitionUsage
>;

/**
 * Record<{@link FilterNameId}, Record<{@link ChannelName}, Set<{@link ChannelSegmentDescriptorId}>>>
 *
 * A record of filter name ids (a combination of the namedFilter name + filter definition name), each
 * containing a record of channel names to a set of unique channel segment descriptor ids
 * (stringified channelSegmentDescriptors).
 */
export type ProcessedItemsCacheRecord = Record<
  FilterNameId,
  Record<ChannelName, Set<ChannelSegmentDescriptorId>>
>;

/**
 * Record<{@link SignalDetectionIdString}, {@link SignalDetectionTypes.SignalDetection}>
 *
 * A record of signal detection ids to signal detections.
 */
export type SignalDetectionsRecord = Record<
  SignalDetectionIdString,
  SignalDetectionTypes.SignalDetection
>;

/**
 * Record<{@link EventId}, {@link EventTypes.Event}>
 *
 * A record of event names to an event.
 */
export type EventsRecord = Record<EventId, EventTypes.Event>;

/**
 * Record<{@link SignalDetectionIdString}, {@link EventTypes.AssociationConflict}>
 *
 * A record of Association Conflicts.
 */
export type AssociationConflictRecord = Record<SignalDetectionIdString, AssociationConflict>;

/**
 * Record<{@link ChannelName}, Record<{@link QcSegmentId}, {@link QcSegment}>>
 *
 * A record of channel names containing a record of qc segment id's to qc segments.
 */
export type QcSegmentRecord = Record<ChannelName, Record<QcSegmentId, QcSegment>>;

/**
 * Used by {@link FilterAssociation} in conjunction
 * with {@link FilterDefinitionAssociationsObject}.
 * Represents the id/startTime of a ChannelSegment that was filtered using a
 * particular {@link FilterTypes.FilterDefinition}
 */
export interface WaveformIdentifier {
  channelSegmentId: string;
  startTime: number;
}

/**
 * Used in conjunction with {@link FilterDefinitionAssociationsObject}.
 * Represents a collection of ChannelSegments that were filtered with a
 * particular {@link FilterTypes.FilterDefinition}
 */
export interface FilterAssociation {
  definition: FilterTypes.FilterDefinition;
  waveformIdentifiers: WaveformIdentifier[];
}

/**
 * This object associates {@link WeavessTypes.DataSegment}s to
 * designed {@link FilterTypes.FilterDefinition} objects for the purpose of file exporting.
 */
export interface FilterDefinitionAssociationsObject {
  filterAssociations: FilterAssociation[];
  channelSegments: UiChannelSegment[];
}

/**
 * Record<{@link ChannelSegmentDescriptorId}, {@link SignalDetectionHypothesisId}>
 *
 * A record of channel segment descriptor strings to signal detection hypothesis id
 */
export type ChannelSegmentsToSignalDetectionHypothesisRecord = Record<
  ChannelSegmentDescriptorId,
  SignalDetectionHypothesisId
>;
