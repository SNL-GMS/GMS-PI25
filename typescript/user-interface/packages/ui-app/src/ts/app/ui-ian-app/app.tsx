import { HotkeysProvider, Intent, NonIdealState, Spinner } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { Displays } from '@gms/common-model';
import { IanDisplays } from '@gms/common-model/lib/displays/types';
import { IS_MODE_IAN } from '@gms/common-util';
import { getStore, useHistoryHotKeys, useUiTheme, withReduxProvider } from '@gms/ui-state';
import { useGlobalHotkeyListener } from '@gms/ui-util/lib/ui-util/hot-key-util';
import React from 'react';
import { Provider } from 'react-redux';
import { HashRouter, Route, Routes } from 'react-router-dom';

import { AnalystUiComponents } from '~analyst-ui/';
import { AnalystCommandRegistrar } from '~analyst-ui/commands/analyst-command-registrar';
import { FilterManager } from '~analyst-ui/components/filters/filter-manager';
import { InteractionConsumer } from '~analyst-ui/interactions/interaction-consumer';
import { InteractionProvider } from '~analyst-ui/interactions/interaction-provider';
import { authenticator } from '~app/authentication';
import { useGlobalPreventDefault } from '~app/global-hotkey-hooks';
import { EffectiveNowTimeInitializer } from '~app/initializers/effective-time-now-initializer';
import { UIThemeWrapper } from '~app/initializers/ui-theme-wrapper';
import { LoadingIndicatorWrapper } from '~app/loading-indicator-wrapper';
import { ServiceWorkerController } from '~app/service-worker-controller';
import { ThemedToastContainer } from '~app/themed-toast-container';
import { UniqueComponent } from '~common-ui/components/unique-component';
import { CommonCommandRegistrar } from '~components/common-ui/commands';
import {
  CommandPalette,
  useToggleCommandPaletteVisibilityHotkeys
} from '~components/common-ui/components/command-palette';
import { LoadingScreen } from '~components/loading-screen';
import { LoginScreen } from '~components/login-screen';
import { ProtectedRoute } from '~components/protected-route';
import { GoldenLayoutContext } from '~components/workspace/components/golden-layout/types';
import { ReduxWorkspaceContainer as Workspace } from '~components/workspace/workspace-container';
import { GMS_DISABLE_KEYCLOAK_AUTH } from '~env';

import { createPopoutComponent } from '../create-popout-component';
import { glContextData } from './golden-layout-config';

const GMS = 'GMS';
const store = getStore();

/**
 * Wraps the component route.
 * Provides the required context providers to the component.
 *
 * @param Component the component route
 * @param props the props passed down from the route to the component
 * @param suppressPopinIcon true to force suppress the golden-layout popin icon
 */
function wrap(Component: any, name: string | undefined, props: any, suppressPopinIcon = false) {
  function WrappedComponent(p) {
    const [uiTheme] = useUiTheme();
    useGlobalHotkeyListener();
    useGlobalPreventDefault();

    useHistoryHotKeys();
    useToggleCommandPaletteVisibilityHotkeys();

    const unique = name ? (
      <UniqueComponent name={name}>
        {/* eslint-disable-next-line react/jsx-props-no-spreading */}
        <Component {...p} />
      </UniqueComponent>
    ) : (
      // eslint-disable-next-line react/jsx-props-no-spreading
      <Component {...p} />
    );

    return (
      <ServiceWorkerController>
        <HotkeysProvider>
          <>
            <UIThemeWrapper key={uiTheme.name}>
              <CommandPalette>
                <LoadingIndicatorWrapper />
                {unique}
                <InteractionProvider>
                  <InteractionConsumer />
                </InteractionProvider>
                <CommonCommandRegistrar />
                <AnalystCommandRegistrar />
              </CommandPalette>
              <ThemedToastContainer />
            </UIThemeWrapper>
            <EffectiveNowTimeInitializer />
            <FilterManager />
          </>
        </HotkeysProvider>
      </ServiceWorkerController>
    );
  }

  return createPopoutComponent(withReduxProvider(WrappedComponent), props, suppressPopinIcon);
}

const WrappedWorkflow: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.Workflow, IanDisplays.WORKFLOW, props);
const WrappedEvents: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.Events, IanDisplays.EVENTS, props);
const WrappedFilters: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.Filters, IanDisplays.FILTERS, props);
const WrappedWaveform: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.Waveform, IanDisplays.WAVEFORM, props);
const WrappedMap: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.IANMap, IanDisplays.MAP, props);
const WrappedSignalDetections: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.SignalDetections, IanDisplays.SIGNAL_DETECTIONS, props);
const WrappedStationProperties: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.StationProperties, IanDisplays.STATION_PROPERTIES, props);
const WrappedWorkspace: React.FC<unknown> = props => wrap(Workspace, undefined, props, true);
const WrappedAzimuthSlowness: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.AzimuthSlowness, IanDisplays.AZIMUTH_SLOWNESS, props);
const WrappedHistory: React.FC<unknown> = props =>
  wrap(AnalystUiComponents.History, IanDisplays.HISTORY, props);

export function App(): React.ReactElement {
  const redirectRoute = GMS_DISABLE_KEYCLOAK_AUTH ? (
    <>
      <Route path="/login/:redirectUrl" element={<LoginScreen authenticator={authenticator} />} />
      <Route path="/login" element={<LoginScreen authenticator={authenticator} />} />
    </>
  ) : undefined;

  return !IS_MODE_IAN ? (
    <NonIdealState
      icon={IconNames.ERROR}
      action={<Spinner intent={Intent.DANGER} />}
      title="Invalid settings"
      description="Not configured for IAN mode - Please check settings"
    />
  ) : (
    <Provider store={store}>
      <HotkeysProvider>
        <HashRouter>
          {
            // ! CAUTION: when changing the route paths
            // The route paths must match the `golden-layout` component name for popout windows
            // For example, the component name `my-route` must have the route path of `my-route`
          }
          {
            // For performance use `render` which accepts a functional component
            // that won't get unnecessarily remounted like with component.
          }
          <Routes>
            {
              // Authentication
              // React router dropped support for optional params
              // so we need separate routes for login and login with a redirect
            }
            {redirectRoute}
            <Route
              path="/loading"
              element={
                <ProtectedRoute redirectPath="loading" title={`${GMS}: Loading...`}>
                  <LoadingScreen />
                </ProtectedRoute>
              }
            />
            {
              // Individual ProtectedRoutes
            }
            <Route
              path={`/${IanDisplays.WORKFLOW}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.WORKFLOW}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.WORKFLOW, `${GMS}: `)}
                >
                  <WrappedWorkflow />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.EVENTS}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.EVENTS}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.EVENTS, `${GMS}: `)}
                >
                  <WrappedEvents />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.FILTERS}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.FILTERS}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.FILTERS, `${GMS}: `)}
                >
                  <WrappedFilters />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.WAVEFORM}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.WAVEFORM}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.WAVEFORM, `${GMS}: `)}
                >
                  <WrappedWaveform />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.MAP}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.MAP}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.MAP, `${GMS}: `)}
                >
                  <WrappedMap />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.SIGNAL_DETECTIONS}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.SIGNAL_DETECTIONS}
                  title={Displays.toDisplayTitle(
                    Displays.IanDisplays.SIGNAL_DETECTIONS,
                    `${GMS}: `
                  )}
                >
                  <WrappedSignalDetections />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.AZIMUTH_SLOWNESS}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.AZIMUTH_SLOWNESS}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.AZIMUTH_SLOWNESS, `${GMS}: `)}
                >
                  <WrappedAzimuthSlowness />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.STATION_PROPERTIES}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.STATION_PROPERTIES}
                  title={Displays.toDisplayTitle(
                    Displays.IanDisplays.STATION_PROPERTIES,
                    `${GMS}: `
                  )}
                >
                  <WrappedStationProperties />
                </ProtectedRoute>
              }
            />
            <Route
              path={`/${IanDisplays.HISTORY}`}
              element={
                <ProtectedRoute
                  redirectPath={IanDisplays.HISTORY}
                  title={Displays.toDisplayTitle(Displays.IanDisplays.HISTORY, `${GMS}: `)}
                >
                  <WrappedHistory />
                </ProtectedRoute>
              }
            />
            {
              // Workspace
            }
            <Route
              path="*"
              element={
                <ProtectedRoute title={`${GMS}: App Workspace`}>
                  <GoldenLayoutContext.Provider value={glContextData()}>
                    <WrappedWorkspace />
                  </GoldenLayoutContext.Provider>
                </ProtectedRoute>
              }
            />
          </Routes>
        </HashRouter>
      </HotkeysProvider>
    </Provider>
  );
}
