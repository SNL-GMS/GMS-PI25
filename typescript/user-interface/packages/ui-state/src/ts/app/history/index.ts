export * from './actions';
export * from './history-hooks';
export * from './history-keyboard-hooks';
export * from './history-selectors';
export * from './history-slice';
export * from './middleware';
export * from './reducers';
