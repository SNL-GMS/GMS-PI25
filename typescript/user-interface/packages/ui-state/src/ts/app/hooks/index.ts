export * from './action-target-hooks';
export * from './beamforming-template-hooks';
export * from './channel-hooks';
export * from './channel-segment-hooks';
export * from './data-acquisition-hooks';
export * from './event-hooks';
export * from './event-manager-hooks';
export * from './filter-hooks';
export * from './fk-channel-segment-hooks';
export * from './keyboard-shortcuts-hooks';
export * from './operational-time-period-configuration-hooks';
export * from './processing-analyst-configuration-hooks';
export * from './processing-mask-hooks';
export * from './processing-station-hooks';
export * from './qc-segment-hooks';
export * from './query-util-hooks';
export * from './react-redux-hooks';
export * from './signal-detection-hooks';
export * from './signal-enhancement-configuration-hooks';
export * from './station-definition-hooks';
export * from './system-event-gateway-hooks';
export * from './update-signal-detection-hooks';
export * from './user-manager-hooks';
export * from './user-session-hooks';
export * from './waveform-hooks';
export * from './workflow-hooks';
