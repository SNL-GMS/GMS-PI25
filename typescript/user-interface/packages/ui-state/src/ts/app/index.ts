export * from './api';
export * from './history';
export * from './hooks';
export * from './query';
export * from './redux-provider';
export * from './sanitization';
export * from './state';
export * from './store';
export * from './subscription';
export * from './util';
