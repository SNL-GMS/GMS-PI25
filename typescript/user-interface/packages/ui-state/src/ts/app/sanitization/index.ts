export * from './config';
export * from './ui-state-help';
