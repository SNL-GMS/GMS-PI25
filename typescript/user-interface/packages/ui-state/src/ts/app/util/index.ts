export * as ChannelFactory from './channel-factory';
export * from './channel-segment-util';
export * from './data-segment-util';
export * from './ui-waveform-masking-util';
