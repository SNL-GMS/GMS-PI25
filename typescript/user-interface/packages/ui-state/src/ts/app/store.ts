import { IS_NODE_ENV_PRODUCTION, NODE_ENV } from '@gms/common-util';
import { getElectron, getElectronEnhancer, UILogger } from '@gms/ui-util';
import { configureStore } from '@reduxjs/toolkit';
import { createLogger } from 'redux-logger';
import type { Config } from 'redux-state-sync';
import {
  createStateSyncMiddleware,
  initStateWithPrevTab,
  withReduxStateSync
} from 'redux-state-sync';

import {
  eventManagerApiSlice,
  processingConfigurationApiSlice,
  processingStationApiSlice,
  signalEnhancementConfigurationApiSlice,
  sohAceiApiSlice,
  ssamControlApiSlice,
  stationDefinitionSlice,
  systemEventGatewayApiSlice,
  systemMessageDefinitionApiSlice,
  userManagerApiSlice,
  workflowApiSlice
} from './api';
import { eventConflictMiddleware } from './api/data/event/event-conflict-middleware';
import {
  eventRemovedMiddleware,
  eventUndoRedoMiddleware,
  historyAddAction,
  historyMiddleware,
  signalDetectionRemovedMiddleware,
  undoRedoMiddleware
} from './history';
import { reducer } from './reducer';
import { reduxDevToolsGlobalConfig } from './sanitization';
import { buildDenyList } from './sanitization/deny-list';
import { buildActionSanitizer, sanitizeState } from './sanitization/sanitizers';
import { getDevToolsConfigLogs } from './sanitization/util';
import { userSessionActions } from './state';
import {
  establishWsConnection,
  registerConnectionStatusCallback
} from './subscription/subscription';

const logger = UILogger.create('GMS_LOG_REDUX_STORE', process.env.GMS_LOG_REDUX_STORE || 'info');

const reduxStateSyncConfig: Config = {
  blacklist: [
    // to not sync the history add to prevent duplicates since the middleware is running in all windows
    historyAddAction.type
  ]
};

const buildStore = () => {
  const DISABLE_REDUX_STATE_SYNC = reduxDevToolsGlobalConfig.GMS_DISABLE_REDUX_STATE_SYNC.enabled;

  logger.info(
    `Configuring Redux store with the following properties: \n` +
      `NODE_ENV:${NODE_ENV}\n${getDevToolsConfigLogs()} `
  );

  const store = configureStore({
    reducer: !DISABLE_REDUX_STATE_SYNC ? withReduxStateSync(reducer) : reducer,
    devTools:
      !IS_NODE_ENV_PRODUCTION && !reduxDevToolsGlobalConfig.GMS_DISABLE_REDUX_DEV_TOOLS.enabled
        ? {
            trace: reduxDevToolsGlobalConfig.GMS_ENABLE_REDUX_TRACE.enabled,
            actionsDenylist: buildDenyList(),
            actionSanitizer: buildActionSanitizer(),
            stateSanitizer: sanitizeState
          }
        : false,
    middleware: getDefaultMiddleware => {
      const middlewares = getDefaultMiddleware({
        thunk: true,
        immutableCheck: reduxDevToolsGlobalConfig.GMS_ENABLE_REDUX_IMMUTABLE_CHECK.enabled,
        serializableCheck: reduxDevToolsGlobalConfig.GMS_ENABLE_REDUX_SERIALIZABLE_CHECK.enabled
      })
        // add the listener middleware to the store
        // NOTE: since this can receive actions with functions inside it should go before the serialize check middleware
        // ! the order of the middle matters is important!
        .prepend(undoRedoMiddleware.middleware)
        .prepend(eventUndoRedoMiddleware.middleware)
        .prepend(historyMiddleware.middleware)
        .prepend(eventRemovedMiddleware.middleware)
        .prepend(signalDetectionRemovedMiddleware.middleware)
        .prepend(eventConflictMiddleware.middleware)
        // ! custom middlewares must be prepended before this line
        .concat(systemEventGatewayApiSlice.middleware)
        .concat(eventManagerApiSlice.middleware)
        .concat(processingConfigurationApiSlice.middleware)
        .concat(processingStationApiSlice.middleware)
        .concat(sohAceiApiSlice.middleware)
        .concat(ssamControlApiSlice.middleware)
        .concat(signalEnhancementConfigurationApiSlice.middleware)
        .concat(stationDefinitionSlice.middleware)
        .concat(systemMessageDefinitionApiSlice.middleware)
        .concat(userManagerApiSlice.middleware)
        .concat(workflowApiSlice.middleware);

      if (!DISABLE_REDUX_STATE_SYNC) {
        middlewares.push(createStateSyncMiddleware(reduxStateSyncConfig));
      }

      // ! the logger should always be the last middleware added
      // enable the Redux logger only if `GMS_ENABLE_REDUX_LOGGER` is set to true
      if (reduxDevToolsGlobalConfig.GMS_ENABLE_REDUX_LOGGER.enabled) {
        const reduxLogger = createLogger({
          collapsed: true,
          duration: true,
          timestamp: false,
          level: 'info',
          logger: console,
          logErrors: true,
          diff: false
        });
        middlewares.push(reduxLogger);
      }
      return middlewares;
    },
    enhancers:
      getElectron() && getElectronEnhancer()
        ? [
            // must be placed after the enhancers which dispatch
            // their own actions such as redux-thunk or redux-saga
            getElectronEnhancer()({
              dispatchProxy: a => store.dispatch(a)
            })
          ]
        : []
  });

  // initialize state using any previous state from other tabs
  if (!DISABLE_REDUX_STATE_SYNC) {
    initStateWithPrevTab(store);
  }

  // store is created connect to System Event gateway and set callback to set connection status
  establishWsConnection();
  const updateConnectionStatus = (connected: boolean): void => {
    if (store) {
      store.dispatch(userSessionActions.setConnected(connected));
    }
  };
  registerConnectionStatusCallback(updateConnectionStatus, store.getState().app.userSession);

  return store;
};

export type ReduxStoreType = ReturnType<typeof buildStore>;

export interface GMSWindow extends Window {
  ReduxStore: ReduxStoreType;
}

const getGmsStore = (): ReturnType<typeof buildStore> => {
  const gmsWindow = (window as unknown) as GMSWindow;
  // ! ensure that only one Redux store instance is created
  if (gmsWindow.ReduxStore == null) {
    gmsWindow.ReduxStore = buildStore();
  }
  return gmsWindow.ReduxStore;
};

/**
 * Creates the Redux application store, which is stored on the window object in order to avoid any chance for
 * multiple stores to be created (if multiple copies of the module are included in the bundle, for example).
 *
 * @returns the Redux store for the application
 */
export const getStore = (): ReduxStoreType => getGmsStore();

/**
 * Returns the dispatch function of the Redux store/
 *
 * @returns the Redux dispatch function
 */
export const getDispatch = () => getGmsStore().dispatch;

/**
 * The application state.
 * Infer the `AppState` types from the store itself
 */
export type AppState = ReturnType<typeof reducer>;

/**
 * The keys of the application state.
 */
export type AppStateKeys = keyof AppState;

/**
 * The application dispatched (typed).
 * Infer the `AppDispatch` types from the store itself
 */
export type AppDispatch = ReturnType<typeof getDispatch>;
