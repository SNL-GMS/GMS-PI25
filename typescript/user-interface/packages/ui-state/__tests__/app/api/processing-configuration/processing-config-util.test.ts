import type { HotkeyConfiguration } from '@gms/common-model/lib/ui-configuration/types';

import { getStore } from '../../../../src/ts/app';
import {
  buildHotkeyConfigArray,
  getUiTheme
} from '../../../../src/ts/app/api/processing-configuration';

const store = getStore();
describe('Processing Configuration Util', () => {
  it('can retrieve ui theme configuration', () => {
    // eslint-disable-next-line @typescript-eslint/unbound-method
    expect(getUiTheme(store.getState)).toMatchSnapshot();
  });
  it('can convert a HotkeyConfiguration to a blueprint HotkeyConfig', () => {
    const shortcutConfig: HotkeyConfiguration = {
      description: 'Test description',
      combos: ['ctrl+x', 'alt+x'],
      helpText: 'help text',

      categories: ['category', 'category2']
    };

    const mockCallback = jest.fn();

    const result = buildHotkeyConfigArray(shortcutConfig, mockCallback);

    // Global and disabled default to false
    expect(result[0].global).toBeFalsy();
    expect(result[0].disabled).toBeFalsy();

    // Records are built for both combinations
    expect(result).toHaveLength(2);
    expect(result[0].combo).toEqual('ctrl+x');
    expect(result[1].combo).toEqual('alt+x');

    // Disabled can be set to true
    const result2 = buildHotkeyConfigArray(shortcutConfig, mockCallback, undefined, true);
    expect(result2[0].global).toBeFalsy();
    expect(result2[0].disabled).toBeTruthy();

    // Global can be set to true
    const result3 = buildHotkeyConfigArray(shortcutConfig, mockCallback, undefined, false, true);
    expect(result3[0].global).toBeTruthy();
    expect(result3[0].disabled).toBeFalsy();
  });
});
