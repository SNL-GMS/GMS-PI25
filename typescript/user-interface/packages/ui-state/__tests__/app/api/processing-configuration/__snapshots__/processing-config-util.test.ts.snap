// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processing Configuration Util can retrieve ui theme configuration 1`] = `
{
  "colors": {
    "completeEventSDColor": "#62D96B",
    "completeEventSDHoverColor": "#BBFFBC",
    "conflict": "#FF0000",
    "deletedEventColor": "#ff6347",
    "deletedSdColor": "#ff6347",
    "fkNeedsReview": "#EBFF00",
    "fkSelection": "#00FFFF",
    "gmsActionTarget": "#ffffff",
    "gmsBackground": "#182026",
    "gmsMain": "#f5f8fa",
    "gmsMainInverted": "#10161a",
    "gmsSelection": "#1589d1",
    "gmsTableSelection": "#f5f8fa",
    "mapStationDefault": "#6F6E74",
    "mapVisibleStation": "#D9822B",
    "openEventSDColor": "#C87619",
    "otherEventSDColor": "#FFFFFF",
    "popover": {
      "background": "#30404d",
      "gridBackground": "#3d4b58",
      "header": "#30404d",
      "input": "303d46",
      "invalidInput": "#1c2127",
    },
    "predictionSDColor": "#C58C1B",
    "qcMaskColors": {
      "analystDefined": "#EB06C8",
      "dataAuthentication": "#8A57FF",
      "longTerm": "#0E9B96",
      "processingMask": "#F87C2E",
      "rejected": "#FF0000",
      "stationSOH": "#B58400",
      "unprocessed": "#FFFFFF",
      "waveform": "#00E22B",
    },
    "rejectedEventColor": "#ab83f1",
    "unassociatedSDColor": "#C23030",
    "waveformDimPercent": 0.75,
    "waveformFilterLabel": "#F6F7F9",
    "waveformMaskLabel": "#EB06C8",
    "waveformRaw": "#3FA6DA",
    "weavessOutOfBounds": "#10161a",
  },
  "display": {
    "edgeEventOpacity": 0.5,
    "edgeSDOpacity": 0.2,
    "predictionSDOpacity": 0.1,
  },
  "isDarkMode": true,
  "name": "GMS Default (dark)",
}
`;
